File: internal/subagent/activity.go
package subagent

import (
	"context"

	"nowhere-agent/internal/agent"
)

// Activity is a subagent progress signal surfaced to the run panel and, for
// streamed output, to the chat thread. It is a UI hint only — never part of the
// conversation the model sees, and never persisted (delivered as a live-only
// content event).
type Activity struct {
	AgentType string `json:"agentType"`
	Depth     int    `json:"depth"`
	Phase     string `json:"phase"`          // "start" | "stream" | "tool" | "result" | "interrupted" | "done" | "error"
	Tool      string `json:"tool,omitempty"` // tool name, for phase "tool"/"result"
	// ToolCallID ties the signal to the spawn_agent tool call that launched this
	// child, so the UI can nest output under the right card when several
	// subagents run in parallel.
	ToolCallID string `json:"toolCallId,omitempty"`
	// Kind classifies a Phase "stream" payload: "text" | "thinking".
	Kind string `json:"kind,omitempty"`
	// Text is one streamed delta of the child's own output (Phase "stream").
	Text string `json:"text,omitempty"`
	// SubToolCallID is the child's own tool call's id (Phase "tool"/"result"),
	// so the UI can render that tool invocation (and match its result) exactly
	// like a top-level call.
	SubToolCallID string `json:"subToolCallId,omitempty"`
	// Args is the child tool call's input (Phase "tool"); Result its output and
	// IsError its error flag (Phase "result").
	Args    any    `json:"args,omitempty"`
	Result  string `json:"result,omitempty"`
	IsError bool   `json:"isError,omitempty"`
}

// Sink receives subagent activity for forwarding to the UI. The run worker
// installs one via WithSink; a nil sink (no runtime, e.g. tests/dev) means
// subagents run black-box.
type Sink func(Activity)

type sinkKey struct{}

// WithSink returns ctx carrying a subagent activity sink. Because the sink rides
// in the run context, nested subagents at any depth report to the same run
// stream without extra plumbing.
func WithSink(ctx context.Context, s Sink) context.Context {
	return context.WithValue(ctx, sinkKey{}, s)
}

func sinkFrom(ctx context.Context) Sink {
	s, _ := ctx.Value(sinkKey{}).(Sink)
	return s
}

// activityEmitter forwards a child loop's events to the run's activity sink as
// compact Activity signals. It stands in for the discard emitter when a sink is
// present. The child's collapsed result still reaches the parent as the tool
// result; this emitter only mirrors progress + a live copy of the child's
// streamed text/thinking so the UI can show the subagent working in real time.
type activityEmitter struct {
	sink       Sink
	agentType  string
	depth      int
	toolCallID string
	// interrupted records that the child ended awaiting client interactions
	// (approval/ask_user/client_tool). The loop still emits KindDone after its
	// interrupt frames; that trailing "done" is suppressed so the UI doesn't
	// show a gated child as finished.
	interrupted bool
}

func (e *activityEmitter) Emit(_ context.Context, kind agent.EventKind, payload any) error {
	if e.sink == nil {
		return nil
	}
	base := Activity{AgentType: e.agentType, Depth: e.depth, ToolCallID: e.toolCallID}
	switch kind {
	case agent.KindText:
		if s, ok := payload.(string); ok && s != "" {
			base.Phase, base.Kind, base.Text = "stream", "text", s
			e.sink(base)
		}
	case agent.KindThinking:
		if s, ok := payload.(string); ok && s != "" {
			base.Phase, base.Kind, base.Text = "stream", "thinking", s
			e.sink(base)
		}
	case agent.KindToolUse:
		base.Phase = "tool"
		if m, ok := payload.(map[string]any); ok {
			base.Tool, _ = m["name"].(string)
			base.SubToolCallID, _ = m["id"].(string)
			base.Args = m["input"]
		}
		e.sink(base)
	case agent.KindToolResult:
		base.Phase = "result"
		if m, ok := payload.(map[string]any); ok {
			base.Tool, _ = m["name"].(string)
			base.SubToolCallID, _ = m["tool_use_id"].(string)
			base.Result, _ = m["content"].(string)
			base.IsError, _ = m["is_error"].(bool)
		}
		e.sink(base)
	case agent.KindInterrupt:
		// The child hit a gate it cannot resolve (subagents can't suspend for
		// human input). Surface it so the UI shows the child stalled on an
		// approval/question rather than silently finishing.
		e.interrupted = true
		base.Phase = "interrupted"
		if g, ok := payload.(agent.Interaction); ok {
			base.Tool = g.ToolName
			base.SubToolCallID = g.ToolCallID
			base.Kind = g.Kind
		}
		e.sink(base)
	case agent.KindDone:
		if e.interrupted {
			return nil
		}
		base.Phase = "done"
		e.sink(base)
	case agent.KindError:
		base.Phase = "error"
		e.sink(base)
	}
	return nil
}

File: web/src/lib/activity.ts
// A tiny in-app event bus feeding the right-hand panel. The chat thread is the
// source of truth for what the agent did; components there (tool calls, text)
// publish activity here, and the panel's tabs (Runs / Workspace) subscribe.
// Keeping it outside assistant-ui means the panel re-renders independently of
// the high-frequency streaming store.

import { useSyncExternalStore } from "react";

export type ToolActivity = {
  id: string;
  toolName: string;
  argsText: string;
  result?: unknown;
  isError?: boolean;
  status: "running" | "done" | "error";
  /** Epoch ms when the call started. */
  at: number;
};

// A subagent run aggregated from the backend's live `data-subagent` activity
// signals (start → stream/tool/result… → done/error). It is a UI progress hint
// only; the subagent's actual output reaches the model as the spawn_agent tool
// result.
//
// The run is kept as an ordered part list (thinking / text / tool-call) so the
// chat card can render it with the SAME components as a top-level assistant
// message — a subagent's own spawn_agent nesting recurses naturally.
export type SubPart =
  | { kind: "thinking"; text: string }
  | { kind: "text"; text: string }
  | {
      kind: "tool";
      id: string;
      toolName: string;
      argsText: string;
      result?: unknown;
      isError?: boolean;
      status: "running" | "done" | "error";
    };

export type SubagentRun = {
  id: number;
  agentType: string;
  depth: number;
  tools: string[]; // flat tool-name list for the compact Runs-panel row
  status: "running" | "done" | "error";
  at: number;
  /** The spawn_agent tool-call id this child belongs to (links to the chat card). */
  toolCallId?: string;
  /** Ordered content parts of the child's turn(s), streamed live. */
  parts: SubPart[];
};

export type SubagentSignal = {
  agentType: string;
  depth: number;
  phase: "start" | "stream" | "tool" | "result" | "interrupted" | "done" | "error";
  tool?: string;
  toolCallId?: string;
  /** "text" | "thinking" for stream signals; the interaction kind (e.g. "approval") for interrupted. */
  kind?: string;
  text?: string;
  subToolCallId?: string;
  args?: unknown;
  result?: unknown;
  isError?: boolean;
};

export type ActivityState = {
  /** Tool calls in the order they started, updated in place as they finish. */
  tools: ToolActivity[];
  /** Subagent runs spawned during this conversation. */
  subagents: SubagentRun[];
};

let state: ActivityState = { tools: [], subagents: [] };
let subagentSeq = 0;
// epoch bumps on every resetActivity. A gated tool call resumes on a FRESH
// backend run whose re-stream (decision-follow / poll-attach) can still be
// draining when the user switches or starts a new chat — its late report would
// otherwise re-add the just-cleared rows to the new conversation's panel.
// Reports captured against an older epoch are dropped.
let epoch = 0;
const listeners = new Set<() => void>();

function emit() {
  for (const l of listeners) l();
}

function setState(next: ActivityState) {
  state = next;
  emit();
}

// reportToolCall upserts a tool activity by id: called as a call starts and
// again when its result lands, so the Runs tab reflects progress live. `atEpoch`
// is the conversation epoch the reporter is streaming for; a stale report (from
// a run whose conversation was already reset) is ignored.
export function reportToolCall(
  entry: Omit<ToolActivity, "at"> & { at?: number },
  atEpoch: number = epoch,
) {
  if (atEpoch !== epoch) return;
  const id = entry.id || `${entry.toolName}:${entry.at ?? 0}`;
  const existing = state.tools.findIndex((t) => t.id === id);
  const record: ToolActivity = {
    ...entry,
    id,
    at: entry.at ?? Date.now(),
    result: entry.result,
  };
  if (existing >= 0) {
    const tools = state.tools.slice();
    tools[existing] = { ...tools[existing], ...record };
    setState({ ...state, tools });
  } else {
    setState({ ...state, tools: [...state.tools, record] });
  }
}

// reportSubagentActivity folds one live subagent signal into the run's ordered
// part list: `start` opens a run, `stream` appends to the trailing thinking/text
// part, `tool` opens a tool part, `result` fills it, `done`/`error` closes the
// run. Runs are matched by toolCallId when present (parallel subagents), falling
// back to most-recent running run at that depth.
export function reportSubagentActivity(sig: SubagentSignal, atEpoch: number = epoch) {
  if (atEpoch !== epoch) return;
  const runs = state.subagents.slice();
  if (sig.phase === "start") {
    runs.push({
      id: subagentSeq++,
      agentType: sig.agentType,
      depth: sig.depth,
      tools: [],
      status: "running",
      at: Date.now(),
      toolCallId: sig.toolCallId,
      parts: [],
    });
  } else {
    const idx = findRun(runs, sig);
    if (idx < 0) return; // stray signal (start dropped); ignore
    const run = { ...runs[idx], parts: runs[idx].parts.slice(), tools: runs[idx].tools.slice() };
    switch (sig.phase) {
      case "stream":
        if (sig.text) appendStream(run, sig.kind === "thinking" ? "thinking" : "text", sig.text);
        break;
      case "tool":
        if (sig.tool) {
          run.tools.push(sig.tool);
          run.parts.push({
            kind: "tool",
            id: sig.subToolCallId || `${sig.tool}:${run.parts.length}`,
            toolName: sig.tool,
            argsText: sig.args ? JSON.stringify(sig.args) : "",
            status: "running",
          });
        }
        break;
      case "result":
        fillToolResult(run, sig);
        break;
      case "interrupted":
        // The child hit a gate it cannot resolve (subagents can't suspend for
        // human input) and stopped. Show it as failed with the reason, not as
        // still running — the backend suppresses the trailing "done" here.
        run.status = "error";
        closeOpenTools(run);
        run.parts.push({
          kind: "text",
          text: sig.tool
            ? `Stopped: needs ${sig.kind || "approval"} for ${sig.tool}, which cannot be delivered inside a subagent.`
            : "Stopped: waiting on input that cannot be delivered inside a subagent.",
        });
        break;
      case "done":
        run.status = "done";
        closeOpenTools(run);
        break;
      case "error":
        run.status = "error";
        closeOpenTools(run);
        break;
    }
    runs[idx] = run;
  }
  setState({ ...state, subagents: runs.slice(-50) });
}

// appendStream appends a delta to the trailing part of the same kind, starting
// a new part when the kind changes (preserves think/text order like a top turn).
function appendStream(run: SubagentRun, kind: "thinking" | "text", delta: string) {
  const last = run.parts[run.parts.length - 1];
  if (last && last.kind === kind) {
    run.parts[run.parts.length - 1] = { ...last, text: last.text + delta };
  } else {
    run.parts.push({ kind, text: delta });
  }
}

// fillToolResult matches a child tool-result back onto its tool part by id (or
// the latest still-running tool part), marking it done/error with its output.
function fillToolResult(run: SubagentRun, sig: SubagentSignal) {
  for (let i = run.parts.length - 1; i >= 0; i--) {
    const p = run.parts[i];
    if (p.kind !== "tool") continue;
    const idMatch = sig.subToolCallId && p.id === sig.subToolCallId;
    const fallback = !sig.subToolCallId && p.status === "running";
    if (idMatch || fallback) {
      run.parts[i] = {
        ...p,
        result: sig.result,
        isError: sig.isError,
        status: sig.isError ? "error" : "done",
      };
      return;
    }
  }
}

function closeOpenTools(run: SubagentRun) {
  run.parts = run.parts.map((p) =>
    p.kind === "tool" && p.status === "running" ? { ...p, status: "done" } : p,
  );
}

// findRun locates the running subagent a signal belongs to: by toolCallId when
// the signal carries one (parallel spawns), else the most recent running run at
// the same depth (legacy signals without an id).
function findRun(runs: SubagentRun[], sig: SubagentSignal): number {
  if (sig.toolCallId) {
    for (let i = runs.length - 1; i >= 0; i--) {
      if (runs[i].toolCallId === sig.toolCallId) return i;
    }
  }
  for (let i = runs.length - 1; i >= 0; i--) {
    if (runs[i].status === "running" && runs[i].depth === sig.depth) return i;
  }
  return -1;
}

// resetActivity clears per-conversation activity when switching threads so the
// panel doesn't show a previous chat's runs. Bumping the epoch also invalidates
// any report still in flight from the old conversation's streams.
export function resetActivity() {
  epoch++;
  setState({ tools: [], subagents: [] });
}

// activityEpoch exposes the current conversation epoch so a streaming reporter
// (tool-call card, subagent feed) can tag its reports and have them dropped
// once the conversation resets underneath it.
export function activityEpoch(): number {
  return epoch;
}

export function subscribeActivity(cb: () => void): () => void {
  listeners.add(cb);
  return () => listeners.delete(cb);
}

export function getActivity(): ActivityState {
  return state;
}

// useActivity subscribes a component to the activity feed.
export function useActivity(): ActivityState {
  return useSyncExternalStore(subscribeActivity, getActivity);
}
